modules/table/src/main/java/org/apache/ignite/internal/table/distributed/index/IndexBuilder.java

```diff
@@ -0,0 +1,186 @@
+import org.apache.ignite.internal.logger.Loggers;
+import org.apache.ignite.internal.raft.service.RaftGroupService;
+import org.apache.ignite.internal.storage.MvPartitionStorage;
+import org.apache.ignite.internal.storage.index.IndexStorage;
+import org.apache.ignite.internal.table.distributed.command.BuildIndexCommand;
+import org.apache.ignite.internal.thread.NamedThreadFactory;
+import org.apache.ignite.internal.util.IgniteSpinBusyLock;
+import org.apache.ignite.internal.util.IgniteUtils;
+
+/**
+ * Class for managing the building of table indexes.
+ */
+public class IndexBuilder implements ManuallyCloseable {
+    private static final IgniteLogger LOG = Loggers.forClass(IndexBuilder.class);
+
+    private static final int BATCH_SIZE = 100;
+
+    private final ExecutorService executor;
+
+    private final Map<IndexBuildTaskId, IndexBuildTask> indexBuildTaskById = new ConcurrentHashMap<>();
+
+    private final IgniteSpinBusyLock busyLock = new IgniteSpinBusyLock();
+
+    private final AtomicBoolean closeGuard = new AtomicBoolean();
+
+    /**
+     * Constructor.
+     *
+     * @param nodeName Node name.
+     * @param threadCount Number of threads to build indexes.
+     */
+    public IndexBuilder(String nodeName, int threadCount) {
+        executor = new ThreadPoolExecutor(
+                threadCount,
+                threadCount,
+                30,
+                TimeUnit.SECONDS,
+                new LinkedBlockingQueue<>(),
+                NamedThreadFactory.create(nodeName, "build-index", LOG)
+        );
+    }
+
+    /**
+     * Starts building the index if it is not already built or is not yet in progress.
+     *
+     * <p>Index is built in batches using {@link BuildIndexCommand} (via raft), batches are sent sequentially.
+     *
+     * <p>It is expected that the index building is triggered by the leader of the raft group.
+     *
+     * @param tableId Table ID.
+     * @param partitionId Partition ID.
+     * @param indexId Index ID.
+     * @param indexStorage Index storage to build.
+     * @param partitionStorage Multi-versioned partition storage.
+     * @param raftClient Raft client.
+     */
+    // TODO: IGNITE-19498 Perhaps we need to start building the index only once
+    public void startBuildIndex(
+            UUID tableId,
+            int partitionId,
+            UUID indexId,
+            IndexStorage indexStorage,
+            MvPartitionStorage partitionStorage,
+            RaftGroupService raftClient
+    ) {
+        inBusyLock(() -> {
+            if (indexStorage.getNextRowIdToBuild() == null) {
+                return;
+            }
+
+            IndexBuildTaskId taskId = new IndexBuildTaskId(tableId, partitionId, indexId);
+
+            IndexBuildTask newTask = new IndexBuildTask(taskId, indexStorage, partitionStorage, raftClient, executor, busyLock, BATCH_SIZE);
+
+            IndexBuildTask previousTask = indexBuildTaskById.putIfAbsent(taskId, newTask);
```

Comment:
you should use `computeIfAbsent` in order to avoid creating unnecessary task instances

Reply:
think that your suggestion does not quite suit me, because then the code for starting the task and subscribing to the future will be inside the closure, which can lead to a lengthening of the closure execution, although it should be fast, and the deletion of the task can immediately be performed in this closure and we can freeze.

Reply:
ok, I agree